sdk/core/azure-core/src/http/challenge_based_authentication_policy.cpp

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/http/policies/policy.hpp"
+#include <chrono>
+#include <vector>
+
+using Azure::Core::Context;
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::Policies::_internal;
+using namespace Azure::Core::Http::Policies::_detail;
+
+std::unique_ptr<RawResponse> ChallengeBasedAuthenticationPolicy::Send(
+    Request& request,
+    NextHttpPolicy nextPolicy,
+    Context const& context) const
+{
+  request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+  auto result = nextPolicy.Send(request, context);
+
+  if (result->GetStatusCode() == HttpStatusCode::Unauthorized)
+  {
+    auto rawData = result->GetHeaders().find("www-authenticate");
+    if (rawData != result->GetHeaders().end())
+    {
+      ChallengeParameters challenge(rawData->second);
+      Credentials::TokenRequestContext tokenRequestContext;
+
+      tokenRequestContext.TenantId = std::move(challenge.TenantId);
+      tokenRequestContext.AuthorizationUri = std::move(challenge.AuthorizationUri);
+      tokenRequestContext.Scopes = std::move(m_tokenRequestContext.Scopes);
+
+      {
+        std::lock_guard<std::mutex> lock(m_accessTokenMutex);
+        m_accessToken = m_credential->GetToken(tokenRequestContext, context);
+        request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+      }
+    }
+    return nextPolicy.Send(request, context);
```

Comment:
this return should be inside the previous `if`. It should be triggered only when `www-authenticate` is found in the response headers. Otherwise, we should just return the previous `result`.

As an alternative, consider doing:

```cpp
  auto result = nextPolicy.Send(request, context);
  auto const& headers = result->GetHeaders();
  auto const& rawDataHeader = headers.find("www-authenticate");
  // Only re-run when Unauthozied AND www-authenticate header in the response
  if (result->GetStatusCode() == HttpStatusCode::Unauthorized && rawData != headers.end())
  {
      ChallengeParameters challenge(rawData->second);
      Credentials::TokenRequestContext tokenRequestContext;

      tokenRequestContext.TenantId = std::move(challenge.TenantId);
      tokenRequestContext.AuthorizationUri = std::move(challenge.AuthorizationUri);
      tokenRequestContext.Scopes = std::move(m_tokenRequestContext.Scopes);

      {
        std::lock_guard<std::mutex> lock(m_accessTokenMutex);
        m_accessToken = m_credential->GetToken(tokenRequestContext, context);
        request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
      }
       return nextPolicy.Send(request, context);
   }

   // Did not re-send on any other case
   return result;
    
```